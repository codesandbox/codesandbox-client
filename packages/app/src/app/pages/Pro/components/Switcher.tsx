import React from 'react';

import { TeamAvatar } from 'app/components/TeamAvatar';
import styled from 'styled-components';
import {
  Badge,
  Stack,
  Menu,
  Icon,
  Element,
  Text,
} from '@codesandbox/components';
import {
  CurrentTeamInfoFragmentFragment,
  SubscriptionType,
  TeamFragmentDashboardFragment,
  TeamMemberAuthorization,
} from 'app/graphql/types';
import { useWorkspaceSubscription } from 'app/hooks/useWorkspaceSubscription';
import { useActions, useAppState } from 'app/overmind';

export const Switcher: React.FC<{
  workspaces: TeamFragmentDashboardFragment[];
  setActiveTeam: (payload?: { id: string }) => Promise<void>;
  activeTeamInfo: CurrentTeamInfoFragmentFragment;
  personalWorkspaceId: string;
}> = ({ workspaces, setActiveTeam, personalWorkspaceId, activeTeamInfo }) => {
  const { isFree } = useWorkspaceSubscription();
  const { openCreateTeamModal } = useActions();
  const { user } = useAppState();

  return (
    <Menu>
      <MenuTrigger>
        <Text>{activeTeamInfo.name}</Text>
        {isFree ? <Badge variant="trial">Free</Badge> : null}
      </MenuTrigger>

      <MenuList>
        <Element css={{ minWidth: 290, maxHeight: 600, overflow: 'auto' }}>
          <Text
            size={12}
            lineHeight="16px"
            css={{ display: 'block', padding: '8px 24px' }}
          >
            Select team to upgrade
          </Text>

          {workspaces.map(workspace => {
            if (!workspace) return null;

            const userAuthorization = workspace.userAuthorizations.find(
              ({ userId }) => userId === user?.id
            );

            const isBillingManager =
              userAuthorization?.authorization ===
                TeamMemberAuthorization.Admin || userAuthorization?.teamManager;

            const isTrialEligible =
              workspace.id !== personalWorkspaceId &&
              workspace.subscription === null;

            const isPro = [
              SubscriptionType.TeamPro,
              SubscriptionType.PersonalPro,
            ].includes(workspace.subscription?.type);

            const disabled = !(isBillingManager || isTrialEligible);

            return (
              <MenuItem
                onSelect={() => setActiveTeam(workspace)}
                key={workspace.id}
                disabled={disabled}
                style={{ opacity: disabled ? 0.5 : 1, padding: 0 }}
              >
                <Stack css={{ padding: '12px 24px' }} align="center" gap={4}>
                  <TeamAvatar
                    size="small"
                    avatar={workspace.avatarUrl}
                    name={workspace.name}
                    css={{ border: '1px solid #5f5f5d', borderRadius: '2px' }}
                  />

                  <Stack
                    align="center"
                    justify="space-between"
                    gap={3}
                    css={{ flex: 1 }}
                  >
                    <Text
                      size={4}
                      css={{
                        maxWidth: '180px',
                        overflow: 'hidden',
                        textOverflow: 'ellipsis',
                      }}
                    >
                      {workspace.id === personalWorkspaceId
                        ? 'Personal'
                        : workspace.name}
                    </Text>
                    {!isPro && <Badge variant="trial">Free</Badge>}
                  </Stack>
                </Stack>
              </MenuItem>
            );
          })}

          <Element marginY={2} css={{ borderTop: '1px solid #5f5f5d' }} />

          <MenuItem onSelect={openCreateTeamModal}>
            <Stack
              gap={4}
              align="center"
              as={Element}
              marginY={2}
              paddingY={3}
              paddingX={6}
            >
              <Stack
                justify="center"
                align="center"
                css={{
                  width: '24px', // To match Avatar
                  height: '24px', // To match Avatar
                  border: '1px solid #5f5f5d',
                  borderRadius: '2px',
                }}
              >
                <Icon name="plus" size={16} css={{ display: 'block' }} />
              </Stack>

              <Text size={4}>Create a new team</Text>
            </Stack>
          </MenuItem>
        </Element>
      </MenuList>
    </Menu>
  );
};

const MenuList = styled(Menu.List)`
  &[data-reach-menu-list][data-component='MenuList'] {
    background-color: #373737;
    margin-top: 10px;
    margin-left: -2px;
    border: 0;
    border-radius: 0;
    overflow: visible;
  }
`;

const MenuItem = styled(Menu.Item)`
  &[data-reach-menu-item][data-component='MenuItem'] {
    padding: 0;

    &[data-selected] {
      background-color: #484848;
    }
  }
`;

const StyledMenuButton = styled(Menu.Button)`
  height: auto;
  padding: 8px 12px;
  color: #ebebeb;

  // Turned off the transition as it was causing a weird
  // flicker when focussing the button
  transition: none;

  &:hover,
  &:focus {
    color: #ffffff;
    background-color: #252525;
  }

  &:focus {
    outline: 1px solid #ac9cff;
  }
`;

const MenuTrigger = ({ children }: { children: React.ReactNode }) => {
  return (
    <Stack gap={2} as={StyledMenuButton}>
      {children}
      <Icon css={{ color: '#fff' }} name="chevronDown" size={8} />
    </Stack>
  );
};
