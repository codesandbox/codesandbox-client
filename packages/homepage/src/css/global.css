@font-face {
  font-family: 'TWKEverett';
  src: url('../assets/fonts/TWKEverett-Medium-web.woff') format('woff'),
    url('../assets/fonts/TWKEverett-Medium-web.ttf') format('ttf');
}

@import url('https://fonts.googleapis.com/css2?family=Inter:wght@400;500&display=swap');

:root {
  color-scheme: dark;
}

html,
body {
  text-rendering: optimizeLegibility;
  font-smooth: always;
  -webkit-tap-highlight-color: transparent;
  -webkit-touch-callout: none;
  min-height: 100%;
  -webkit-text-size-adjust: 100%;
  height: 100%;
  background: #030303;
  font-size: 16px;
  color: white;
  font-family: Inter, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto,
    Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue';
  /* overflow-x: hidden; */
  -webkit-font-smoothing: antialiased;
}

@media screen and (-webkit-max-device-pixel-ratio: 1.3),
  (max-resolution: 120dpi) {
  html,
  body {
    -webkit-font-smoothing: antialiased;
    -moz-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
    font-smoothing: antialiased;
  }
}

* {
  transition: color, background-color 200ms ease;
}

input {
  -webkit-font-smoothing: antialiased;
  -moz-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  font-smooth: always;
  -webkit-tap-highlight-color: transparent;
  -webkit-touch-callout: none;
}

.disable-hover,
.disable-hover * {
  pointer-events: none !important;
}

iframe {
  margin-bottom: 0;
}

a {
  color: #fff;
}

.gatsby-resp-image-image {
  box-shadow: none !important;
}

.gatsby-resp-image-background-image {
  background-image: none !important;
}

twitter-widget {
  margin-left: auto;
  margin-right: auto;
}

@media (prefers-reduced-motion: reduce) {
  *,
  *::before,
  *::after {
    animation-duration: 0.001s !important;
    transition-duration: 0.001s !important;
  }
}

@media (prefers-reduced-motion: reduce) {
  * {
    animation-duration: 0.01ms !important;
    animation-iteration-count: 1 !important;
    transition-duration: 0.01ms !important;
    scroll-behavior: auto !important;
  }
}
