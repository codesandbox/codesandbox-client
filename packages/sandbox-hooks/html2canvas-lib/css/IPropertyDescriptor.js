export var PropertyDescriptorParsingType;
(function (PropertyDescriptorParsingType) {
    PropertyDescriptorParsingType[PropertyDescriptorParsingType["VALUE"] = 0] = "VALUE";
    PropertyDescriptorParsingType[PropertyDescriptorParsingType["LIST"] = 1] = "LIST";
    PropertyDescriptorParsingType[PropertyDescriptorParsingType["IDENT_VALUE"] = 2] = "IDENT_VALUE";
    PropertyDescriptorParsingType[PropertyDescriptorParsingType["TYPE_VALUE"] = 3] = "TYPE_VALUE";
    PropertyDescriptorParsingType[PropertyDescriptorParsingType["TOKEN_VALUE"] = 4] = "TOKEN_VALUE";
})(PropertyDescriptorParsingType || (PropertyDescriptorParsingType = {}));
