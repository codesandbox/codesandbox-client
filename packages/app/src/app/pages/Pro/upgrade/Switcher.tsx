import React from 'react';

import { TeamAvatar } from 'app/components/TeamAvatar';
import styled from 'styled-components';
import {
  Stack,
  Avatar,
  Menu,
  Icon,
  Element,
  Text,
  Tooltip,
} from '@codesandbox/components';
import {
  CurrentTeamInfoFragmentFragment,
  SubscriptionType,
  TeamFragmentDashboardFragment,
  TeamMemberAuthorization,
} from 'app/graphql/types';
import { Caption } from './elements';

export const Switcher: React.FC<{
  workspaces: TeamFragmentDashboardFragment[];
  setActiveTeam: (payload?: { id: string }) => Promise<void>;
  workspaceType: 'pro' | 'teamPro';
  activeTeamInfo: CurrentTeamInfoFragmentFragment;
  personalWorkspaceId: string;
  userId: string;
  openCreateTeamModal: () => void;
}> = ({
  workspaces,
  workspaceType,
  setActiveTeam,
  personalWorkspaceId,
  activeTeamInfo,
  userId,
  openCreateTeamModal,
}) => {
  if (!workspaces || workspaces.length === 0) return null;

  const members = activeTeamInfo.users.length;
  const memberLabel = `${members} member${members > 1 ? 's' : ''}`;
  const isPersonalWorkspace = workspaceType === 'pro';

  return (
    <Stack
      justify="space-between"
      align="center"
      css={{ fontFamily: "'Inter', sans-serif" }}
    >
      <Menu>
        <Stack as={Menu.Button} css={{ padding: 0, height: 'auto' }}>
          <TeamAvatar
            size="bigger"
            avatar={activeTeamInfo?.avatarUrl}
            name={activeTeamInfo.name}
          />

          <Stack css={{ marginLeft: 24, marginTop: -6 }} direction="vertical">
            <WorkspaceName>
              <span>{activeTeamInfo.name}</span>
            </WorkspaceName>
            <WorkspaceType>
              {isPersonalWorkspace ? 'personal team' : memberLabel}
            </WorkspaceType>
          </Stack>
        </Stack>

        <MenuList>
          <Element css={{ minWidth: 290 }}>
            <Caption
              css={{
                margin: 0,
                height: 40,
                lineHeight: '40px',
                paddingTop: 8,
                paddingLeft: 24,
                color: '#999999',
              }}
            >
              Select team to upgrade
            </Caption>

            {workspaces.map(workspace => {
              if (!workspace) return null;

              const seats = workspace.users.length;
              const seatsLabel = `${seats} member${seats > 1 ? 's' : ''}`;

              const isAdmin =
                workspace.userAuthorizations.find(
                  team => team.userId === userId
                ).authorization === TeamMemberAuthorization.Admin;
              const isPro = [
                SubscriptionType.TeamPro,
                SubscriptionType.PersonalPro,
              ].includes(workspace.subscription?.type);
              const disabled = !isAdmin || isPro;

              return (
                <MenuItem
                  onSelect={() => setActiveTeam(workspace)}
                  key={workspace.id}
                  disabled={disabled}
                  style={{ opacity: disabled ? 0.5 : 1 }}
                >
                  <Stack css={{ padding: '12px 24px' }} align="center">
                    <TeamAvatar
                      size="small"
                      avatar={workspace.avatarUrl}
                      name={workspace.name}
                    />

                    <Stack
                      direction="vertical"
                      css={{ flex: 1, marginLeft: 19 }}
                    >
                      <Stack>
                        <Text size={4}>{workspace.name}</Text>
                        {isPro && <ProBadge>Pro</ProBadge>}
                      </Stack>

                      <Text size={3}>
                        {workspace.id === personalWorkspaceId
                          ? 'personal team'
                          : seatsLabel}
                      </Text>
                    </Stack>

                    {workspace.id === activeTeamInfo.id && (
                      <Icon css={{ marginLeft: 16 }} name="simpleCheck" />
                    )}
                  </Stack>
                </MenuItem>
              );
            })}

            <MenuItem onSelect={openCreateTeamModal}>
              <Stack
                css={{
                  marginTop: 12,
                  padding: 24,
                  borderTop: '1px solid rgba(255, 255, 255, 0.2)',
                }}
                align="center"
              >
                <Stack
                  justify="center"
                  align="center"
                  css={{
                    border: '1px solid #C5C5C5',
                    width: 24,
                    height: 24,
                  }}
                >
                  <Icon name="plus" size={14} />
                </Stack>

                <Text size={4} css={{ marginLeft: 19 }}>
                  Create a new team
                </Text>
              </Stack>
            </MenuItem>
          </Element>
        </MenuList>
      </Menu>

      <Stack
        css={{
          display: 'none',
          '@media (min-width: 720px)': {
            display: 'flex',
          },
        }}
      >
        {!isPersonalWorkspace && (
          <>
            {activeTeamInfo.users.map((user, index) => {
              if (index > (activeTeamInfo.users.length > 4 ? 2 : 3)) {
                return null;
              }

              return (
                <Tooltip label={user.username} key={user.id}>
                  <div>
                    <Avatar css={{ marginLeft: 4 }} user={user} />
                  </div>
                </Tooltip>
              );
            })}

            {members - 3 > 1 && (
              <WorkspaceSeats>
                <span>{members - 3}</span>

                <Dialog>
                  <Stack
                    direction="vertical"
                    gap={3}
                    align="flex-start"
                    css={{
                      overflow: 'auto',
                      height: '100%',
                      padding: '8px 12px',
                    }}
                  >
                    {activeTeamInfo.users.slice(0, members - 3).map(user => {
                      return (
                        <Stack key={user.id} gap={3}>
                          <Avatar user={user} key={user.id} />
                          <Text variant="muted">{user.username}</Text>
                        </Stack>
                      );
                    })}
                  </Stack>
                </Dialog>
              </WorkspaceSeats>
            )}
          </>
        )}
      </Stack>
    </Stack>
  );
};

const Dialog = styled.div`
  position: absolute;
  top: 40px;
  height: 148px;

  background: #373737;

  border: 1px solid #2a2a2a;
  box-sizing: border-box;
  box-shadow: 0px 2px 7px rgba(0, 0, 0, 0.15);

  display: none;

  &:before {
    content: '';
    height: 10px;
    width: 100%;
    position: absolute;
    top: -10px;
    display: block;
  }
`;

const WorkspaceName = styled.p`
  font-family: 'TWKEverett', sans-serif;
  font-style: normal;
  font-weight: 500;

  margin: 0;
  position: relative;
  text-align: left;
  color: #e5e5e5;

  font-size: 24px;

  @media (min-width: 720px) {
    font-size: 32px;
    line-height: 42px;
  }

  & span::-moz-selection {
    -webkit-text-stroke: 1px #e5e5e5;
    color: transparent;
    background: transparent;
  }

  & span::selection {
    -webkit-text-stroke: 1px #e5e5e5;
    color: transparent;
    background: transparent;
  }

  &:after {
    content: '';
    position: absolute;
    right: -20px;
    top: calc(50% - 4px);
    border: 5px solid transparent;
    border-top: 7px solid white;
  }
`;

const WorkspaceType = styled.p`
  font-style: normal;
  font-weight: normal;
  font-size: 16px;
  line-height: 24px;
  margin: 0;
  text-align: left;

  color: #808080;
`;

const WorkspaceSeats = styled.div`
  background: #373737;
  border: 1 solid #2a2a2a;
  width: 32px;
  height: 32px;
  border-radius: 32px;
  display: flex;
  margin-left: 4px;
  position: relative;

  &:hover ${Dialog} {
    display: block;
  }

  span {
    margin: auto;
    font-size: 12px;
  }
`;

const MenuList = styled(Menu.List)`
  &[data-reach-menu-list][data-component='MenuList'] {
    background-color: #373737;
    margin-top: 10px;
    margin-left: -2px;
    border: 0;
    border-radius: 0;
    overflow: visible;
  }
`;

const MenuItem = styled(Menu.Item)`
  &[data-reach-menu-item][data-component='MenuItem'] {
    padding: 0;

    &[data-selected] {
      background-color: #484848;
    }
  }
`;

const ProBadge = styled.p`
  border-radius: 2px;
  background-color: rgba(229, 229, 229, 0.1);
  color: #c5c5c5;

  width: 35px;
  height: 18px;

  text-align: center;
  line-height: 18px;
  font-size: 13px;

  margin: 0 0 0 8px;
`;
