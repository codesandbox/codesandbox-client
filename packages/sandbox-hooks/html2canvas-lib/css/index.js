import { PropertyDescriptorParsingType } from './IPropertyDescriptor';
import { backgroundClip } from './property-descriptors/background-clip';
import { backgroundColor } from './property-descriptors/background-color';
import { backgroundImage } from './property-descriptors/background-image';
import { backgroundOrigin } from './property-descriptors/background-origin';
import { backgroundPosition } from './property-descriptors/background-position';
import { backgroundRepeat } from './property-descriptors/background-repeat';
import { backgroundSize } from './property-descriptors/background-size';
import { borderBottomColor, borderLeftColor, borderRightColor, borderTopColor } from './property-descriptors/border-color';
import { borderBottomLeftRadius, borderBottomRightRadius, borderTopLeftRadius, borderTopRightRadius } from './property-descriptors/border-radius';
import { borderBottomStyle, borderLeftStyle, borderRightStyle, borderTopStyle } from './property-descriptors/border-style';
import { borderBottomWidth, borderLeftWidth, borderRightWidth, borderTopWidth } from './property-descriptors/border-width';
import { color } from './property-descriptors/color';
import { display } from './property-descriptors/display';
import { float, FLOAT } from './property-descriptors/float';
import { letterSpacing } from './property-descriptors/letter-spacing';
import { lineBreak } from './property-descriptors/line-break';
import { lineHeight } from './property-descriptors/line-height';
import { listStyleImage } from './property-descriptors/list-style-image';
import { listStylePosition } from './property-descriptors/list-style-position';
import { listStyleType } from './property-descriptors/list-style-type';
import { marginBottom, marginLeft, marginRight, marginTop } from './property-descriptors/margin';
import { overflow } from './property-descriptors/overflow';
import { overflowWrap } from './property-descriptors/overflow-wrap';
import { paddingBottom, paddingLeft, paddingRight, paddingTop } from './property-descriptors/padding';
import { textAlign } from './property-descriptors/text-align';
import { position, POSITION } from './property-descriptors/position';
import { textShadow } from './property-descriptors/text-shadow';
import { textTransform } from './property-descriptors/text-transform';
import { transform } from './property-descriptors/transform';
import { transformOrigin } from './property-descriptors/transform-origin';
import { visibility, VISIBILITY } from './property-descriptors/visibility';
import { wordBreak } from './property-descriptors/word-break';
import { zIndex } from './property-descriptors/z-index';
import { isIdentToken, Parser } from './syntax/parser';
import { Tokenizer } from './syntax/tokenizer';
import { color as colorType, isTransparent } from './types/color';
import { angle } from './types/angle';
import { image } from './types/image';
import { opacity } from './property-descriptors/opacity';
import { textDecorationColor } from './property-descriptors/text-decoration-color';
import { textDecorationLine } from './property-descriptors/text-decoration-line';
import { isLengthPercentage, ZERO_LENGTH } from './types/length-percentage';
import { fontFamily } from './property-descriptors/font-family';
import { fontSize } from './property-descriptors/font-size';
import { isLength } from './types/length';
import { fontWeight } from './property-descriptors/font-weight';
import { fontVariant } from './property-descriptors/font-variant';
import { fontStyle } from './property-descriptors/font-style';
import { contains } from '../core/bitwise';
import { content } from './property-descriptors/content';
import { counterIncrement } from './property-descriptors/counter-increment';
import { counterReset } from './property-descriptors/counter-reset';
import { quotes } from './property-descriptors/quotes';
import { boxShadow } from './property-descriptors/box-shadow';
export class CSSParsedDeclaration {
    constructor(declaration) {
        this.backgroundClip = parse(backgroundClip, declaration.backgroundClip);
        this.backgroundColor = parse(backgroundColor, declaration.backgroundColor);
        this.backgroundImage = parse(backgroundImage, declaration.backgroundImage);
        this.backgroundOrigin = parse(backgroundOrigin, declaration.backgroundOrigin);
        this.backgroundPosition = parse(backgroundPosition, declaration.backgroundPosition);
        this.backgroundRepeat = parse(backgroundRepeat, declaration.backgroundRepeat);
        this.backgroundSize = parse(backgroundSize, declaration.backgroundSize);
        this.borderTopColor = parse(borderTopColor, declaration.borderTopColor);
        this.borderRightColor = parse(borderRightColor, declaration.borderRightColor);
        this.borderBottomColor = parse(borderBottomColor, declaration.borderBottomColor);
        this.borderLeftColor = parse(borderLeftColor, declaration.borderLeftColor);
        this.borderTopLeftRadius = parse(borderTopLeftRadius, declaration.borderTopLeftRadius);
        this.borderTopRightRadius = parse(borderTopRightRadius, declaration.borderTopRightRadius);
        this.borderBottomRightRadius = parse(borderBottomRightRadius, declaration.borderBottomRightRadius);
        this.borderBottomLeftRadius = parse(borderBottomLeftRadius, declaration.borderBottomLeftRadius);
        this.borderTopStyle = parse(borderTopStyle, declaration.borderTopStyle);
        this.borderRightStyle = parse(borderRightStyle, declaration.borderRightStyle);
        this.borderBottomStyle = parse(borderBottomStyle, declaration.borderBottomStyle);
        this.borderLeftStyle = parse(borderLeftStyle, declaration.borderLeftStyle);
        this.borderTopWidth = parse(borderTopWidth, declaration.borderTopWidth);
        this.borderRightWidth = parse(borderRightWidth, declaration.borderRightWidth);
        this.borderBottomWidth = parse(borderBottomWidth, declaration.borderBottomWidth);
        this.borderLeftWidth = parse(borderLeftWidth, declaration.borderLeftWidth);
        this.boxShadow = parse(boxShadow, declaration.boxShadow);
        this.color = parse(color, declaration.color);
        this.display = parse(display, declaration.display);
        this.float = parse(float, declaration.cssFloat);
        this.fontFamily = parse(fontFamily, declaration.fontFamily);
        this.fontSize = parse(fontSize, declaration.fontSize);
        this.fontStyle = parse(fontStyle, declaration.fontStyle);
        this.fontVariant = parse(fontVariant, declaration.fontVariant);
        this.fontWeight = parse(fontWeight, declaration.fontWeight);
        this.letterSpacing = parse(letterSpacing, declaration.letterSpacing);
        this.lineBreak = parse(lineBreak, declaration.lineBreak);
        this.lineHeight = parse(lineHeight, declaration.lineHeight);
        this.listStyleImage = parse(listStyleImage, declaration.listStyleImage);
        this.listStylePosition = parse(listStylePosition, declaration.listStylePosition);
        this.listStyleType = parse(listStyleType, declaration.listStyleType);
        this.marginTop = parse(marginTop, declaration.marginTop);
        this.marginRight = parse(marginRight, declaration.marginRight);
        this.marginBottom = parse(marginBottom, declaration.marginBottom);
        this.marginLeft = parse(marginLeft, declaration.marginLeft);
        this.opacity = parse(opacity, declaration.opacity);
        const overflowTuple = parse(overflow, declaration.overflow);
        this.overflowX = overflowTuple[0];
        this.overflowY = overflowTuple[overflowTuple.length > 1 ? 1 : 0];
        this.overflowWrap = parse(overflowWrap, declaration.overflowWrap);
        this.paddingTop = parse(paddingTop, declaration.paddingTop);
        this.paddingRight = parse(paddingRight, declaration.paddingRight);
        this.paddingBottom = parse(paddingBottom, declaration.paddingBottom);
        this.paddingLeft = parse(paddingLeft, declaration.paddingLeft);
        this.position = parse(position, declaration.position);
        this.textAlign = parse(textAlign, declaration.textAlign);
        this.textDecorationColor = parse(textDecorationColor, declaration.textDecorationColor || declaration.color);
        this.textDecorationLine = parse(textDecorationLine, declaration.textDecorationLine);
        this.textShadow = parse(textShadow, declaration.textShadow);
        this.textTransform = parse(textTransform, declaration.textTransform);
        this.transform = parse(transform, declaration.transform);
        this.transformOrigin = parse(transformOrigin, declaration.transformOrigin);
        this.visibility = parse(visibility, declaration.visibility);
        this.wordBreak = parse(wordBreak, declaration.wordBreak);
        this.zIndex = parse(zIndex, declaration.zIndex);
    }
    isVisible() {
        return this.display > 0 && this.opacity > 0 && this.visibility === VISIBILITY.VISIBLE;
    }
    isTransparent() {
        return isTransparent(this.backgroundColor);
    }
    isTransformed() {
        return this.transform !== null;
    }
    isPositioned() {
        return this.position !== POSITION.STATIC;
    }
    isPositionedWithZIndex() {
        return this.isPositioned() && !this.zIndex.auto;
    }
    isFloating() {
        return this.float !== FLOAT.NONE;
    }
    isInlineLevel() {
        return (contains(this.display, 4 /* INLINE */) ||
            contains(this.display, 33554432 /* INLINE_BLOCK */) ||
            contains(this.display, 268435456 /* INLINE_FLEX */) ||
            contains(this.display, 536870912 /* INLINE_GRID */) ||
            contains(this.display, 67108864 /* INLINE_LIST_ITEM */) ||
            contains(this.display, 134217728 /* INLINE_TABLE */));
    }
}
export class CSSParsedPseudoDeclaration {
    constructor(declaration) {
        this.content = parse(content, declaration.content);
        this.quotes = parse(quotes, declaration.quotes);
    }
}
export class CSSParsedCounterDeclaration {
    constructor(declaration) {
        this.counterIncrement = parse(counterIncrement, declaration.counterIncrement);
        this.counterReset = parse(counterReset, declaration.counterReset);
    }
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const parse = (descriptor, style) => {
    const tokenizer = new Tokenizer();
    const value = style !== null && typeof style !== 'undefined' ? style.toString() : descriptor.initialValue;
    tokenizer.write(value);
    const parser = new Parser(tokenizer.read());
    switch (descriptor.type) {
        case PropertyDescriptorParsingType.IDENT_VALUE:
            const token = parser.parseComponentValue();
            return descriptor.parse(isIdentToken(token) ? token.value : descriptor.initialValue);
        case PropertyDescriptorParsingType.VALUE:
            return descriptor.parse(parser.parseComponentValue());
        case PropertyDescriptorParsingType.LIST:
            return descriptor.parse(parser.parseComponentValues());
        case PropertyDescriptorParsingType.TOKEN_VALUE:
            return parser.parseComponentValue();
        case PropertyDescriptorParsingType.TYPE_VALUE:
            switch (descriptor.format) {
                case 'angle':
                    return angle.parse(parser.parseComponentValue());
                case 'color':
                    return colorType.parse(parser.parseComponentValue());
                case 'image':
                    return image.parse(parser.parseComponentValue());
                case 'length':
                    const length = parser.parseComponentValue();
                    return isLength(length) ? length : ZERO_LENGTH;
                case 'length-percentage':
                    const value = parser.parseComponentValue();
                    return isLengthPercentage(value) ? value : ZERO_LENGTH;
            }
    }
    throw new Error(`Attempting to parse unsupported css format type ${descriptor.format}`);
};
