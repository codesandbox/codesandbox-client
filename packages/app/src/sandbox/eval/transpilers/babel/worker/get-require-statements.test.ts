import { transform } from 'babel-standalone';
import dynamicImportSyntax from 'babel-plugin-syntax-dynamic-import';

import detective from './plugins/babel-plugin-detective';
import getRequireStatements from './get-require-statements';

describe('get-require-statements', () => {
  function testAst(code) {
    const result = transform(code, {
      presets: ['es2015', 'react'],
      plugins: [
        'transform-async-to-generator',
        'transform-object-rest-spread',
        'transform-class-properties',
        'transform-decorators-legacy',
        dynamicImportSyntax,
        [detective, { source: true, nodes: true }],
      ],
    });

    expect(getRequireStatements(detective.metadata(result))).toMatchSnapshot();
  }

  it('can find simple requires', () => {
    const code = `
      import React from 'react';
      import { render } from 'react-dom';
    `;

    testAst(code);
  });

  it('can find plain requires', () => {
    const code = `
      const react = require('react');
      const reactDom = require('react-dom');
    `;

    testAst(code);
  });

  it('does not find import promises', () => {
    const code = `
      const reactDom = import('react-dom').then(dom => dom.render('a'));
    `;

    testAst(code);
  });

  it('can find dynamic imports', () => {
    const code = `
      const page = import('./' + this.props.page);
      const page2 = require('./page/' + this.props.page2);
    `;

    testAst(code);
  });

  it('can find reexports', () => {
    const code = `
      export * from './Hello';
    `;

    testAst(code);
  });

  it('can work with real life code', () => {
    const code = `
    import React from 'react'
    import { connect } from 'react-redux'
    import { TransitionGroup, Transition } from 'transition-group'
    import universal from 'react-universal-component'
    import isLoading from '../selectors/isLoading'
    import '../css/Switcher.css'

    const Switcher = ({ page, direction, isLoading }) =>
      <TransitionGroup
        className={\`switcher $\{direction}\`}
        duration={500}
        prefix='slide'
      >
        <Transition key={page}>
          <UniversalComponent page={page} isLoading={isLoading} />
        </Transition>
      </TransitionGroup>

    const UniversalComponent = universal(props => import(\`./$\{props.page}\`), {
      minDelay: 500,
      chunkName: props => props.page,
      loading: () => <div className='spinner'><div /></div>,
      error: () => <div className='notFound'>PAGE NOT FOUND - 404</div>
    })

    const mapState = ({ page, direction, ...state }) => ({
      page,
      direction,
      isLoading: isLoading(state)
    })

    export default connect(mapState)(Switcher)
    `;

    testAst(code);
  });
});
