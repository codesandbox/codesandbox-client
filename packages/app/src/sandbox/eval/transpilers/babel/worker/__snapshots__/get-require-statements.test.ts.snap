// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-require-statements can find dynamic imports 1`] = `
Array [
  Object {
    "path": "./page/",
    "type": "glob",
  },
]
`;

exports[`get-require-statements can find plain requires 1`] = `
Array [
  Object {
    "path": "react",
    "type": "direct",
  },
  Object {
    "path": "react-dom",
    "type": "direct",
  },
]
`;

exports[`get-require-statements can find reexports 1`] = `
Array [
  Object {
    "path": "./Hello",
    "type": "direct",
  },
]
`;

exports[`get-require-statements can find simple requires 1`] = `
Array [
  Object {
    "path": "react",
    "type": "direct",
  },
  Object {
    "path": "react-dom",
    "type": "direct",
  },
]
`;

exports[`get-require-statements can work with real life code 1`] = `
Array [
  Object {
    "path": "react",
    "type": "direct",
  },
  Object {
    "path": "react-redux",
    "type": "direct",
  },
  Object {
    "path": "transition-group",
    "type": "direct",
  },
  Object {
    "path": "react-universal-component",
    "type": "direct",
  },
  Object {
    "path": "../selectors/isLoading",
    "type": "direct",
  },
  Object {
    "path": "../css/Switcher.css",
    "type": "direct",
  },
]
`;

exports[`get-require-statements does not find import promises 1`] = `Array []`;
